src/maxwell-web-view.js

```diff
@@ -0,0 +1,242 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ * Author: Juan Pablo Ugarte <ugarte@endlessm.com>
+ *
+ */
+
+(function() {
+
+function throttle (func, limit) {
+    let stamp, id;
+
+    return () => {
+        let self = this;
+        let args = arguments;
+
+        if (!stamp) {
+            func.apply(self, args);
+            stamp = Date.now();
+            return;
+        }
+
+        clearTimeout(id);
+        id = setTimeout(() => {
+            let now = Date.now();
+            if (now - stamp >= limit) {
+                func.apply(self, args);
+                stamp = now;
+            }
+        }, limit - (Date.now() - stamp));
+    }
+}
+
+function update_positions () {
+    let children = window.maxwell.children;
+    let positions = new Array ();
+
+    for (let id in children) {
+        let child = children[id];
+        let rect = child.getBoundingClientRect();
+        let x = rect.x;
+        let y = rect.y;
+
+        /* Bail if position did not changed */
+        if (child.maxwell_position_x === x && child.maxwell_position_y === y)
+            continue;
+
+        /* Update position in cache */
+        child.maxwell_position_x = x;
+        child.maxwell_position_y = y;
+
+        /* Collect new positions */
+        positions.push ({ id: id, x: x, y: y });
+    }
+
+    /* Update all positions in MaxwellWebView at once to reduce messages */
+    window.webkit.messageHandlers.maxwell_update_positions.postMessage(positions);
+    position_timeout_id = null;
+}
+
+/* Limit to 10Hz */
+let update_positions_throttled = throttle(update_positions, 100);
+
+/* We need to update widget positions on scroll and resize events */
+window.addEventListener("scroll", update_positions_throttled, { passive: true });
+window.addEventListener("resize", update_positions_throttled, { passive: true });
+
+/* We also need to update it on any DOM change */
+function document_mutation_handler (mutations) {
+    let children = new Array ();
+
+    for (var mutation of mutations) {
+        if (mutation.type !== 'childList')
+            continue;
+
+        for (let i = 0, len = mutation.addedNodes.length; i < len; i++) {
+            let child = mutation.addedNodes[i];
+
+            if (!child.id || child.tagName !== 'CANVAS' ||
+                !child.classList.contains('GtkWidget'))
+                continue;
+
+            /* Save original display value */
+            child.maxwell_display_value = child.style.display;
+
+            /* Hide all widgets by default */
+            child.style.display = 'none';
```

Comment:
Would be nice to perhaps re-use the visibility helper function you created for this.